lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,363 @@
+	// sftp.Client.Mkdir does not take an os.FileMode, so this can't either
+	Mkdir(path string) error
+	// Chmod sets file permissions
+	Chmod(path string, mode os.FileMode) error
+	// Chtimes sets file access and modification time
+	Chtimes(path string, atime, mtime time.Time) error
+}
+
+// CreateUploadConfig returns a Config ready to upload files
+func CreateUploadConfig(src []string, dst string, opts Options) *Config {
+	c := &Config{
+		srcPaths: src,
+		dstPath:  dst,
+		srcFS:    &localFS{},
+		dstFS:    &remoteFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c
+}
+
+// CreateDownloadConfig returns a Config ready to download files
+func CreateDownloadConfig(src, dst string, opts Options) *Config {
+	c := &Config{
+		srcPaths: []string{src},
+		dstPath:  dst,
+		srcFS:    &remoteFS{},
+		dstFS:    &localFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c
+}
+
+// setDefaults sets default values
+func (c *Config) setDefaults() {
+	logger := c.Log
+	if logger == nil {
+		logger = log.StandardLogger()
+	}
+	c.Log = logger.WithFields(log.Fields{
+		trace.Component: "SFTP",
+		trace.ComponentFields: log.Fields{
+			"SrcPaths":      c.srcPaths,
+			"DstPath":       c.dstPath,
+			"Recursive":     c.opts.Recursive,
+			"PreserveAttrs": c.opts.PreserveAttrs,
+		},
+	})
+}
+
+// TransferFiles transfers files from the configured source paths to the
+// configured destination path over SFTP
+func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	c.initFS(ctx, sftpClient)
+
+	transferErr := c.transfer(ctx)
+	closeErr := sftpClient.Close()
+	if transferErr != nil {
+		return trace.Wrap(transferErr)
+	}
+	if closeErr != nil {
+		return trace.Wrap(closeErr)
+	}
+	return nil
+}
+
+// initFS ensures the source and destination filesystems are ready to transfer
+func (c *Config) initFS(ctx context.Context, client *sftp.Client) {
+	srcFS, ok := c.srcFS.(*remoteFS)
+	if ok {
+		srcFS.c = client
+	}
+	dstFS, ok := c.dstFS.(*remoteFS)
+	if ok {
+		dstFS.c = client
+	}
+	c.srcFS.SetContext(ctx)
+	c.dstFS.SetContext(ctx)
+}
+
+// transfer preforms file transfers
+func (c *Config) transfer(ctx context.Context) error {
+	var dstIsDir bool
+	dstInfo, err := c.dstFS.Stat(c.dstPath)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return trace.Errorf("error accessing %s path %q: %v", c.dstFS.Type(), c.dstPath, err)
+		}
+		// if there are multiple source paths and the destination path
+		// doesn't exist, create it as a directory
+		if len(c.srcPaths) > 1 {
+			if err := c.dstFS.Mkdir(c.dstPath); err != nil {
+				return trace.Wrap(err)
+			}
+			dstIsDir = true
+		}
+	} else if len(c.srcPaths) > 1 && !dstInfo.IsDir() {
+		// if there are multiple source paths, ensure the destination path
+		// is a directory
+		return trace.BadParameter("%s file %q is not a directory, but multiple source files were specified",
```

Comment:
Is this designed to just be advisory? What does the error look like if we get an ENOTDIR later on in `transferFile` or `transferDir`?

Reply:
Someone swapping out a directory for a file after we check but before we start using it as a directory. That sort of error message is fine.